libbeat/_meta/config/logging.reference.yml.tmpl

```diff
@@ -67,3 +67,39 @@ logging.files:
   # Rotate existing logs on startup rather than appending them to the existing
   # file. Defaults to true.
   # rotateonstartup: true
+
+# Some outputs will log raw events on errors like indexing errors in the
+# Elasticsearch output, to prevent logging raw events together with other
+# log messages, a different log file, only for log entries containing raw events,
+# is used. It will use the same level, selectors and all other configurations
+# from the default logger, but it will have it's own file configuration.
+#logging.events:
+  #files:
+    # Configure the path where the logs are written. The default is the logs directory
+    # under the home path (the binary location).
+    #path: /var/log/{{.BeatName}}
+
+    # The name of the files where the logs are written to.
+    #name: {{.BeatName}}-events-data
+
+    # Configure log file size limit. If the limit is reached, log file will be
+    # automatically rotated.
+    #rotateeverybytes: 10485760 # = 10MB
```

Comment:
For standalone Beats, this doubles the disk usage of our log files by default. This is probably ok, perhaps we could make this smaller.

For Elastic Agent, this uses an additional 10 MB of disk for each individual Beat process that is started. There is also no way to configure this in the Elastic Agent. This also moves us away from the current logging model where there is only a single elastic agent log.

I think for agent, it would be better to have the event logger write to stdout/stderr but obviously identify the log as being event data (perhaps using `log.logger`) and have the agent separately direct the event logs into a single merged file.

Essentially I think this event log concept needs to also exist at the agent level with the event data logger to stdout/stderr the way the rest of the logs. 

What we could do as an intermediate step is lower the size limit here by default, since it in theory is only going to get written for ES index errors. Then do the unified agent log implementation as a follow up. 

I think having a unified agent event log would let us eliminate the agent trace log level we introduced and allow us to log event content unconditionally at the debug level again, which is still disabled in the block below:

https://github.com/elastic/beats/blob/b9e706cf5d6fc62cb8d4ae4ff4ad36e9f572fbff/libbeat/publisher/processing/processors.go#L203-L214

Reply:
> Essentially I think this event log concept needs to also exist at the agent level with the event data logger to stdout/stderr the way the rest of the logs.

This would also let us eliminate the diagnostics hook here.

Reply:
I'll reduce the default size for standalone Beats, what about 5MB per file and 5 files maximum? That's about 25MB if there are so many events logged that the rotation reaches it's limit.

> I think for agent, it would be better to have the event logger write to stdout/stderr but obviously identify the log as being event data (perhaps using log.logger) and have the agent separately direct the event logs into a single merged file.

If we log to stdout/stderr, we effectively maintain the current behaviour when running under Elastic-Agent: Beats will mix event data (which can be sensitive data) with it's normal logs and those can easily end up in the same place.

`log.logger` is a keyworkd field defined in [ECS](https://www.elastic.co/guide/en/ecs/current/ecs-log.html#field-log-logger) as the name of the logger, we use it quite freely in Beats and there are no standards for it within Beats, making it a fragile way of differentiate the log types. Also this field is the same in both loggers, which makes it a easier to correlate the logs from both files.

Using `log.logger` would require the Elastic-Agent to have some knowledge of Beats loggers and which ones contain event/sensitive data, which also seems fragile to me.

I like the idea of still using a file, even, under Elastic-Agent because it "physically" separates both types of log and does not require the Elastic-Agent to know details about Beats logs. The diagnostics hook looks to me like a good alternative for Beats to fully manage those files and still have them in the diagnostics. If the size gRPC payload becomes an issue, we could think about having a way for process under Elastic-Agent to tell it some extra files to be added to the diagnostics bundle (that can even help for logging metrics into a separate file).

> Essentially I think this event log concept needs to also exist at the agent level with the event data logger to stdout/stderr the way the rest of the logs.

If that is the case, I believe the best option would be to add another field to [ECS](https://www.elastic.co/guide/en/ecs/current/ecs-log.html#ecs-log) that indicates the type of the log, e.g: `default`, `sensitive`, `metrics`. That way it is very clear they need to be handled in different ways and even have different rotation/retention strategies.

Reply:
> Beats will mix event data (which can be sensitive data) with it's normal logs and those can easily end up in the same place.

IMO preventing this is always in the hands of the developer. Everything depends on the person writing the code knowing which logs can be sensitive and using the appropriate technique (redaction, using the sensitive logger, etc). 

> If that is the case, I believe the best option would be to add another field to [ECS](https://www.elastic.co/guide/en/ecs/current/ecs-log.html#ecs-log) that indicates the type of the log, e.g: default, sensitive, metrics. That way it is very clear they need to be handled in different ways and even have different rotation/retention strategies.

We don't need a new ECS field to start, the contract would purely be between agent and the sub-processes it runs. Eventually this could be added to ECS, but we don't need to start with modifying ECS. IMO adding to ECS probably doesn't have a point because we should never be ingesting these events to Elasticsearch to begin with, so having a standard field name and mapping there is pointless. 

We don't have to use `log.logger`. We can define a new field, like `log.type` that does not collide with anything currently in ECS https://www.elastic.co/guide/en/ecs/current/ecs-log.html or SemConv https://opentelemetry.io/docs/specs/semconv/general/logs/

Our current approach is too Beat centric. We have not solved this problem in general. Essentially there are two things to address:
1. Support the concept of diverting logs from the stdout/stderr stream to different files. Metrics is a good example of this.
2. Allow agent sub-processes other than Beats to write event data to disk for root cause analysis in a location that is not sent to Fleet (i.e. the existing elastic-agent-*.ndjson files).

Reply:
If we centralize writing potentially sensitive logs through agent this also gives us more opportunities to fix them. For example we could have agent redact them on the fly in the future, or route them to a special index/ingest pipeline that does that. 

We ideally have the sub-processes/Beats annotate which logs are potentially sensitive, but since mistakes are always possible a future approach that can compensate for that as a platform feature is desirable.

Reply:
@cmacknz I added a `logger.type` (3d389d9475) and added the option to log to a file or stderr.

I also renamed everything to "sensitive logger" instead of "events logger" to focus more on sensitive information rather than only events (f8253aaaba). The docs are also updated.